File: csharp/Practice01/src/MyDriver/Program.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace MyDriver
{
    class Program
    {
        static void Main(string[] args)
        {
            var driver = new MyDriver("jeffz://server:12345");

            try
            {
                driver.AddQuery(1);
                driver.AddQuery(2);
                driver.AddQuery(3);
            }
            catch
            {
                Console.WriteLine("Error occurred when add query.");
                Environment.Exit(1);
            }

            new Thread(() => ReceiveData(driver)).Start();
        }

        private static void ReceiveData(MyDriver driver)
        {
            try
            {
                while (true)
                {
                    var data = driver.Receive();
                    if (data == null)
                    {
                        Console.WriteLine("Closed");
                        break;
                    }
                    else
                    {
                        Console.WriteLine(data);
                    }
                }
            }
            catch (MyDriverException)
            {
                Console.WriteLine("Error occurred when receive data.");
            }
        }
    }
}

File: java/Practice01/src/myDriver/Program.java
package myDriver;

public class Program {

    /**
     * @param args
     */
    public static void main(String[] args) {
        
        final MyDriver driver = new MyDriver("jeffz://server:12345");

        try {
        	driver.connect();
            driver.addQuery(1);
            driver.addQuery(2);
            driver.addQuery(3);
        }
        catch (MyDriverException ex) {
            driver.close();
            System.out.println("Error occurred when connect or add query.");
            System.exit(1);
        }

        new Thread(new Runnable() {

            @Override
            public void run() {
                receiveData(driver);
            }
            
        }).start();
    }

    private static void receiveData(MyDriver driver) {
        try {
            while (true) {
                MyData data = driver.receive();
                if (data == null) {
                    System.out.println("Closed");
                    break;
                }
                else {
                    System.out.println(data);
                }
            }
        }
        catch (MyDriverException ex) {
            driver.close();
            System.out.println("Error occurred when receive data.");
        }
    }
}
